internal/observability/observability.go

```diff
@@ -0,0 +1,97 @@
+package observability
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/sourcegraph/sourcegraph/internal/logging"
+	"github.com/sourcegraph/sourcegraph/internal/metrics"
+	"github.com/sourcegraph/sourcegraph/internal/trace"
+)
+
+// EndObservationFn is the shape of the function returned by PrepObservation and should be
+// invoked within a defer directly before the observed function returns.
+type EndObservationFn func(
+	// The number of things processed.
+	count float64,
+	// Fields to log after the operation is performed.
+	logFields ...log.Field,
+)
+
+// PrepObservation prepares the necessary timers, loggers, and metrics to observe an
+// operation. This returns a decorated context, which should be used in place of the input
+// context in the observed operation, and ah EndObservationFn function. This function should
+// be invoked  on defer. If your function does not process a variable number of items and the
+// counting metric counts invocations, the method should be deferred as follows:
+//
+//     func observedFoo(ctx context.Context) (err error) {
+//         ctx, endObservation := PrepObservation(ctx, &err, logger, metrics, tracer, "TraceName", "log-name")
+//         defer endObservation(1)
```

Comment:
For some reason, I think use `With` and `Finish` is more intuitive if I'm going to use these functions with context:
```suggestion
//         ctx, finish := WithObservation(ctx, &err, logger, metrics, tracer, "TraceName", "log-name")
//         defer finish(1)
```

Reply:
Updated.